File: 101/scala/src/main/scala/Palindrome.scala
/**
 * Check if some numbers are palindromic, check if some Strings are palindromic, then find the longest palindrome
 * in some Strings.
 */
object Palindrome {
  /**
   * Find the longest palindrome in a String.
   *
   * @param str the String to search.
   * @return the longest palindrome in the String.
   */
  def longestPalindrome(str: String): String = {
    if(isPalindromicStr(str)) {
      return str
    }

    val len = str.length
    var substrLen = len

    while(substrLen > 1) {
      var substrIndex = 0
      val maxSubstrIndex = len - substrLen + 1

      while(substrIndex <= maxSubstrIndex) {
        val substr: String = str.substring(substrIndex, substrLen + substrIndex - 1)

        if(isPalindromicStr(substr)) {
          return substr
        }

        substrIndex += 1
      }

      substrLen -= 1
    }

    str.charAt(0).toString
  }

  /**
   * Check if a string is a palindrome.
   *
   * @param str the String to check.
   * @return true if the String is a plaindrome, false if not.
   */
  def isPalindromicStr(str: String): Boolean = {
    val len: Int = str.length
    val mid: Int = len / 2
    var i: Int = 0

    while(i < mid) {
      if(str.charAt(i) != str.charAt(len - (1 + i))) {
        return false
      }

      i += 1
    }

    true
  }

  /**
   * Check if a number is palindromic.
   *
   * @param num the number to check.
   * @return true if the number is a palindrome, false if not.
   */
  def isPalindromic(num: Long): Boolean = {
    reverse(num) == num
  }

  /**
   * Reverse a number.
   *
   * @param num the number to reverse.
   * @return the reversed number.
   */
  def reverse(num: Long): Long = {
    var reverse: Long = 0
    var operation: Long = num

    while(operation > 0) {
      reverse = (reverse * 10) + operation % 10
      operation /= 10
    }

    reverse
  }
}

File: 101/ts/src/palindrome.ts
/**
 * Check if a number is a palindrome.
 *
 * @param {number} num the number to check.
 * @returns {boolean} true if the number is a palindrome, false if not.
 */
export const isPalindrome = (num: number): boolean => num === reverse(num);

/**
 * Check if a string is a palindrome.
 *
 * @param {string} str the string to check.
 * @returns {boolean} true if the string is a palindrome, false if not.
 */
export const isPalindromeStr = (str: string): boolean => {
  const mid = Math.floor(str.length / 2);
  const len = str.length;
  let i = 0;

  while (i < mid) {
    if (str.charAt(i) !== str.charAt(len - i - 1)) {
      return false;
    }
    i += 1;
  }

  return true;
};

/**
 * Find the longest palindrome in a string.
 *
 * @param {string} str the string to check.
 * @returns {string } the longest palindrome in the string.
 */
export const longestPalindrome = (str: string): string => {
  const len = str.length;
  let substrLen = len;

  while (substrLen > 0) {
    let idx = 0;

    while (idx < len - substrLen) {
      const subStr = str.substring(idx, idx + substrLen);
      if (isPalindromeStr(subStr)) {
        return subStr;
      }
      idx += 1;
    }
    substrLen -= 1;
  }

  return str.charAt(0).toString();
};

/**
 * Reverse a number.
 *
 * @param {number} num the number to reverse.
 * @returns {number} the reversed number.
 */
const reverse = (num: number): number => {
  let reverse = 0;
  let op = num;

  while (op > 0) {
    reverse = reverse * 10 + (op % 10);
    op = Math.floor(op / 10);
  }

  return reverse;
};
